docs/architecture/adr-047-handling-evidence-from-light-client.md

```diff
@@ -5,210 +5,186 @@
-the advantage of having a full state. It remains possible in future discussions to have the
-light client form the evidence types itself. The other minor downsides apart from the
-load will be that around half the evidence produced by the light client will be invalid and
-that, in the event of modified logic, it's easier and expected that the full node be up 
-to date than the light clients.
-
-## Decision
-
-When two headers have different hashes, the light client must first verify via 
-bisection that the signed header provided by the witness is also valid.
-
-It then collects all the headers that were fetched as part of bisection into two traces. 
-One containing the primary's headers and the other containing the witness's headers. 
-The light client sends the trace to the opposite provider (i.e. the primary's trace
-to the witness and the witness's trace to the primary) as the light client is 
-incapable of deciding which one is malicious. 
-
-Assuming one of the two providers is honest, that full node will then take the trace
-and extract out the relevant evidence and propogate it until it ends up on chain.
-
-*NOTE: We do not send evidence to other connected peers*
-
-*NOTE: The light client halts then and does not verify with any other witness*
-
-## Detailed Design
-
-The traces will have the following data structure: 
+A previously discussed approach to handling evidence was to pass all the data that the
+light client had witnessed when it had observed diverging headers for the full node to
+process.This was known as a light trace and had the following structure:
 
 ```go
 type ConflictingHeadersTrace struct {
   Headers []*types.SignedHeader
 }
 ```
 
-When a full node receives a `ConflictingHeadersTrace`, it should
-a) validate it b) figure out if malicious behaviour is obvious (immediately
-slashable) or run the amnesia protocol.
+This approach has the advantage of not requiring as much processing on the light
+client side in the event that an attack happens. Although, this is not a significant
+difference as the light client would in any case have to validate all the headers
+from both witness and primary. Using traces would consume a large amount of bandwidth
+and adds a DDOS vector to the full node.
 
-### Validating headers
 
-Check headers are valid (`ValidateBasic`), are in order of ascending height, and do not exceed the `MaxTraceSize`.
+## Decision
 
-### Finding Block Bifurcation
+The light client will be divided into two components: a `Verifier` (either sequential or
+skipping) and a `Detector` (see [Informal's Detector](https://github.com/informalsystems/tendermint-rs/blob/master/docs/spec/lightclient/detection/detection.md))
+. The detector will take the trace of headers from the primary and check it against all
+witnesses. For a witness with a diverging header, the detector will first verify the header
+by bisecting through all the heights defined by the trace that the primary provided. If valid,
```

Comment:
Isn't it possible that the witness header can't be verified using the trace provided by the primary, and that we might also need to run a separate bisection trace on the witness?